projects/2-state-machines/chern.ipynb

```diff
@@ -0,0 +1,492 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "ab = \"a\" + \"b\"\n",
+    "print(ab)\n",
+    "abc = ab + \"c\"\n",
+    "print(abc)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can combine lists the same way:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "lst_abc = [\"a\",\"b\"] + [\"c\"]\n",
+    "print(lst_abc)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "All that's left is enumerating the valid paths of a given length and assembling those into strings. The skeleton below is a _recursive_ function that calls itself repeatedly, branching out according to each option. You can imagine that the call to `sample()` is the root of a tree of calls to `sample2`, and the paths through that tree are paths through the state machine---the path itself is \"stored\" in the third argument to `sample2`. In other words:\n",
+    "\n",
+    "* To get the possible paths of length `K` from a state machine `SM`:\n",
+    "    * Initialize `S` to `SM`'s current state.\n",
+    "    * Initialize the `result` to `[]`\n",
+    "    * For each available out-edge at `S`:\n",
+    "        * Advance `SM` along that edge\n",
+    "        * Get the possible paths of length `K-1` from `SM` and add them to the result\n",
+    "        * Put `SM` back into state `S` (so we can advance it along the next available out edge)\n",
+    "    * Yield `result`\n",
+    "\n",
+    "It's up to you to figure out how to find the possible paths of length 0. Hint: How many such paths can there be for a given call to `sample2(sm, 0, sofar)`? What aspects of the state machine's status does it depend on?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# Starter function that calls sample2() with an empty string as argument. This is just to \"accumulate\" paths into\n",
+    "# that argument without tracking a bunch of data externally.\n",
+    "def sample(sm, length):\n",
+    "    return sample2(sm, length, \"\")\n",
+    "\n",
+    "# sample2 samples paths of length \"length\" from the state machine sm as of its current state.\n",
+    "# sofar is the path assembled so far.\n",
+    "# sample2 returns a list of paths.\n",
+    "def sample2(sm, length, sofar):\n",
+    "    if length == 0:\n",
+    "        # TODO: What path(s) should be returned? When?\n",
+    "        if sm.is_terminal:\n",
+    "            return [sofar]\n",
+    "        else:\n",
+    "            return[]\n",
+    "    state = sm.state\n",
+    "    result = []\n",
```

Comment:
Your behavior when `length == 0` is exactly right, but it looks like you missed implementing the loop here.

Let's consider the case where the state `s` of state machine `sm` has two out-edges: "x" and ".". In this case, the possible paths starting from this state start with either "x" or ".". If they start with an "x", we must have taken the edge along "x"; let `s2` be the state that `sm` will be in after calling `sm.advance("x")`, and we can now sample every path starting from s2. The length limit prevents us from looping forever. Once we're done finding the paths assuming we went along "x", we can reset `sm` back to state `s`, advance along "." , and find the possible paths from _that_ successor state.

So, you want a loop over the out edges of `s`, and within that loop you need to:

1. Advance `sm` along `s` (to get it into state `s2`)
2. Sample (`sample2()`) the paths of `sm` of length `length-1` starting from `s2` (what do you pass as the `sofar` argument? And what do you do with the result of this sampling?)
3. Reset `sm` to state `s` for the next iteration of the loop.

Please give it a try! It will help a lot with the planning assignment.